parser-typechecker/src/Unison/Server/Backend.hs

```diff
@@ -810,9 +828,53 @@ renderDoc ppe width rt codebase r = do
         Nothing -> pure ()
       pure $ r <&> Term.amap (const mempty)
 
-    decls (Reference.DerivedId r) = fmap (DD.amap (const ())) <$> lift (Codebase.getTypeDeclaration codebase r)
+    decls (Reference.DerivedId r) = fmap (DD.amap (const ())) <$> Codebase.getTypeDeclaration codebase r
     decls _ = pure Nothing
 
+docsInBranchToHtmlFiles
+  :: Var v
+  => Rt.Runtime v
+  -> Codebase IO v Ann
+  -> Branch IO
+  -> FilePath
+  -> IO ()
+docsInBranchToHtmlFiles runtime codebase currentBranch directory = do
+  let deepTermRefs = (toList . R.range . Branch.deepTerms . Branch.head) currentBranch >>= toList
+  docRefs <- filterM (isDoc codebase) deepTermRefs
+  hqLength <- Codebase.hashLength codebase
+  let printNames = getCurrentPrettyNames (AllNames Path.empty) currentBranch
+  let width = defaultWidth
+  let ppe = PPE.fromNamesDecl hqLength printNames
+  docs <- concat <$> for docRefs (renderDoc ppe width runtime codebase . Referent.toReference)
+  liftIO $ traverse_ (renderDocToHtmlFile directory) docs
+
+  where
+    cleanPath :: FilePath -> FilePath
+    cleanPath filePath =
+      filePath <&> \case
+        '#' -> '@'
+        c   -> c
+
+    docFilePath :: FilePath -> HashQualifiedName -> FilePath
+    docFilePath directory rawHqn =
+      let
+        (path, docName) =
+          fromRight (error "Could not parse doc name") $ PathParse.parseHQSplit $ Text.unpack rawHqn
+
+        directoryPath =
+          directory </> joinPath (map NameSegment.toString $ Path.toList path)
+
+        docFileName =
+          cleanPath $ HQ'.toString docName <> ".html"
+
+      in directoryPath </> docFileName
+
+    renderDocToHtmlFile :: FilePath -> (HashQualifiedName, UnisonHash, Doc.Doc) -> IO ()
+    renderDocToHtmlFile directory (docName, _, doc) =
+      let
+        docFilePath' = docFilePath directory docName
+      in
+      Lucid.renderToFile docFilePath' (DocHtml.toHtml doc)
```

Comment:
If Lucid doesn't let you write to files outside the current directory (kind of an odd limitation...), I'd probably just call `Lucid.renderBS`

https://hackage.haskell.org/package/lucid-2.10.0/docs/Lucid.html#v:renderBS

And then use this to write that to any file you want

https://hackage.haskell.org/package/bytestring-0.11.1.0/docs/src/Data.ByteString.Lazy.html#writeFile

Reply:
@pchiusano I think its probably and overall good security wise, but we should handle the error better.